### DISCUSSION TITLE
NEW ENDPOINT

### DISCUSSION THREAD

**Discussion by lintangtimur** [lintangtimur/ovoid#46] [Q&A] [answered]
Sepertinya ada mekanisme baru saat login. Ini muncul jika anda melakukan login di perangkat yang belum pernah untuk login.

### STEP 1
POST /v3/user/accounts/otp HTTP/1.1

Request
```json
{
  "channel_code": "ovo_android",
  "device_id": "87586011-0127-3469-aa2a-6016a68d45f6",
  "msisdn": "+6212345678901",
  "otp": {
    "locale": "ID",
    "sms_hash": "m9mj4ctIVR8"
  }
}
```
`sms_hash` boleh diganti, nohp saya samarkan.

Response
```json
{
  "response_code": "OV00000",
  "response_message": "Success",
  "data": {
    "otp": {
      "otp_ref_id": "4d5e1669-dae2-48bf-b0e7-3030ebbb16f6",
      "type": "LOGIN",
      "expires_at": 1618766220,
      "retry_remaining": 4,
      "cooldown_period": 60,
      "length": 4,
      "reff_type": "OTP"
    }
  }
}
```

### STEP 2
POST /v3/user/accounts/otp/validation HTTP/1.1
Request
```json
{
  "channel_code": "ovo_android",
  "device_id": "87586011-0127-3469-aa2a-6016a68d45f6",
  "msisdn": "+62xxx",
  "otp": {
    "otp": "1448",
    "otp_ref_id": "4d5e1669-dae2-48bf-b0e7-3030ebbb16f6",
    "type": "LOGIN"
  }
}
```
Response
```json
{
  "response_code": "OV00000",
  "response_message": "Success",
  "data": {
    "otp": {
      "otp_ref_id": "4d5e1669-dae2-48bf-b0e7-3030ebbb16f6",
      "type": "LOGIN",
      "expires_at": 1618765116,
      "otp_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2RlaGFzaCI6Ik56QmxOelJpTnpNNE1EZ3dORFZqWTJGaU9USXdaR1E1WmpZMU5qaGxaakkiLCJyYW5kb20iOiJNakEyTXprME1EUTAiLCJ2ZXJzaW9xxxxxxxxxxxx"
    },
    "registration_origin": {
      "full_name": "SXXXXXXXXX",
      "registration_origin_name": "OVO",
      "email": "linXXXXXXXXXm"
    }
  }
}
````
### STEP 3
Mendapatkan public keys
GET v3/user/public_keys

### STEP 4
POST https://agw.ovo.id/v3/user/accounts/login
Request Body
```json
{
   "channel_code":"ovo_android",
   "credentials":{
      "otp_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2RlaGFzaCI6IllqWXpZek15WmpFeFpqVXlOREEzWm1FeU1HVTJNakJsWWpFME56QXlZVFUiLCJyYW5kb20iOiJNalEwTWpVeU5qTTUiLCJ2ZXJzaW9uIjoxfQ.ap8bYmC05673QA2dTmhHw7L4SLsfpEXb_VG1c53Sxxxxxx",
      "password":{
         "format":"rsa",
         "value":"QsP7bC/kLduJyBoM6gT/8sasabiwnOlt9j8JvbIs4Z/eXQU7ZPYsF3NkBgJGkHQziitktdinvZG8\nBwXpH6h2BvbtNvIa/eBY2x8TUqqNmBWoK3fj2ESNR58VxdJy+kS99WyXqs4SXdKFKBS+Y2N87+Md\nXqrDaxFPZ62JVKQ/uQ8xTNVnFT+zDV6s4i8GqWgKOUSnfknyPRUN4QVHUy4KaEerxOTtmvR2qRLg\nrKUNkFo6RAHS2K59P2UbSLP4OW06gEAJ+502pafHZwdEWZb78WyQLPH+Iym7TC0+ZA0Th71ghX/F\nRo73NuXDkqLUjzkKWL3/Y8N4Y4oEljPwhh+hVg==\n"
      }
   },
   "device_id":"87586011-0127-3469xxxxxxxxxx",
   "msisdn":"+6282xxxxxxxx",
   "push_notification_id":"fs-DYcGaRbKERLhF4hkQ92:APA91bEjjUFzzFvadIKtdrrqsyrGH26xLRR5-Oyym2l9Ybv0O1cnvqA14ghuTbXz0ogazN-Kw6iGxW2klakANBaVXoFCLrT4hWJJ5FCGOz2o5bGE7RX6XpxndNkcxnqpWat449vBvYSa"
}
```
Response
```json
{
   "response_code":"OV00000",
   "response_message":"Success",
   "data":{
      "auth":{
         "access_token":"eyJhbGxxxxxxciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpU2Vjb25kcyI6NjA0ODAwMDAwLCJjcmVhdGVUaW1lIjoxNjIzNDE3NjYyOTU0LCJzZWNyZXQiOiJ5aWVuL1FhNHphTkl3RXdsOUxreXRYUVFQYzNrZkxFS29iTUcxNkM0a080c09RcXhMTzBlbytuRlZSYzMrZXRvUFh0T1VrSi9TMVdBZXpFemJLUWdBVjZYT0VKc2dQWEx4WmpiK0g4VXc4R3NGZGh5S1NDYWowN1JLdEFIdDJTL2RJUkF2amhoeSt5bjZVSXhUYVN6UHlHYXplQlVLOTNqVFFBZTdlQ0x3RW5MWnliZm5SRXRoMFFPUEN6KzFIUDlGSE44Q0duQ1VmRm5vZkhnSDNaN083c2VrSmJRZkdFN1dLZ2RQZUZRdXJSV3hjQ3NHL3IyclQrc1FncVdWU3dEdXRjTkp5ejFaeWhpMkl0L1RHWVBmNXUvSGxJQjdkQmRGMHBGcTdZT3pKeFluRHJ1YW4rMGNNbkJVZldFWnZwY0h6TmJjYzMzaFhHaHlIYnpySEpLbGs3dzRkRlhTNmk5MExLVzZFTGpQaE09In0.a75AvydQ9aTgluJwial4YBKom0PoGJxwdMiKryo5ruzCObWcH56mBChuwKLBHwdUTgGRXbwe07CQPa3i55rtuUEJwOwl6l4Uv3-BEqhUP0oRoNqwdyH1qzaeldQ_77toqL0HG6wFcEU0WnVwVtfQuFoV8GrMSJSCVTwHhxzbKo7VDnsTve8a96tE9o1XFQ0vQywoqn16FR8Dame6ptSRouqJTdsKMW3OAWLOLs0q-3Vpa2AP5IjN-y9v-EMzF4NkRUPhNPmCN91JudIqj-0Eif9q5Vf0JFClDIfB53QVC3nLgGliul5EnhxN70butsL1lBSx7LDPpDF1pN-RerKRCQ",
         "token_type":"Bearer",
         "expires_in":1623504062,
         "refresh_token":"eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpU2Vjb25kcyI6NjA0ODAwMDAwLCJjcmVhdGVUaW1lIjoxNjIzNDE3NjYyOTU0LCJzZWNyZXQiOiJ5aWVuL1FhNHphTkl3RXdsOUxreXRYUVFQYzNrZkxFS29iTUcxNkM0a080c09RcXhMTzBlbytuRlZSYzMrZXRvUFh0T1VrSi9TMVdBZXpFemJLUWdBVjZYT0VKc2dQWEx4WmpiK0g4VXc4R3NGZGh5S1NDYWowN1JLdEFIdDJTL2RJUkF2amhoeSt5bjZVSXhUYVN6UHlHYXplQlVLOTNqVFFBZTdlQ0x3RW5MWnliZm5SRXRoMFFPUEN6KzFIUDlGSE44Q0duQ1VmRm5vZkhnSDNaN083c2VrSmJRZkdFN1dLZ2RQZUZRdXJSV3hjQ3NHL3IyclQrc1FncVdWU3dEdXRjTkp5ejFaeWhpMkl0L1RHWVBmNXUvSGxJQjdkQmRGMHBGcTdZT3pKeFluRHJ1YW4rMGNNbkJVZldFWnZwY0h6TmJjYzMzaFhHaHlIYnpySEpLbGs3dzRkRlhTNmk5MExLVzZFTGpQaE09In0.a75AvydQ9aTgluJwial4YBKom0PoGJxwdMiKryo5ruzCObWcH56mBChuwKLBHwdUTgGRXbwe07CQPa3i55rtuUEJwOwl6l4Uv3-BEqhUP0oRoNqwdyH1qzaeldQ_77toqL0HG6wFcEU0WnVwVtfQuFoV8GrMSJSCVTwHhxzbKo7VDnsTve8a96tE9o1XFQ0vQywoqn16FR8Dame6ptSRouqJTdsKMW3OAWLOLs0q-3Vpa2AP5IjN-y9v-EMzF4NkRUPhNPmCN91JudIqj-0Eif9q5Vf0JFClDIfB53QVC3nLgGliul5EnhxN70butsL1lBSx7LDPpDF1pN-RerKRCQ",
         "scope":"user"
      },
      "account":{
         "ovo_id":"008103xxxxxxx",
         "msisdn":"08213xxxxxxxx",
         "account_status":"REGISTERED",
         "kyc_status":"",
         "authentication":"PROVIDER_OTP_PASSWORD_ENCRYPTED",
         "push_notification_id":"fs-DYcGaRbxxxxxxxxxxA91bEjjUFzzFvadIKtdrrqsyrGH26xLRR5-Oyym2l9Ybv0O1cnvqA14ghuTbXz0ogazN-Kw6iGxW2klakANBaVXoFCLrT4hWJJ5FCGOz2o5bGE7RX6XpxndNkcxnqpWat449vBvYSa"
      }
   }
}
```

**Comment by cvkjm** [comment]
![image](https://user-images.githubusercontent.com/75083007/121718377-7224ea00-cb0c-11eb-9718-f831dbd2e402.png)

Not found nih

**Reply by lintangtimur** [reply]
yang v3 base url nya https://agw.ovo.id/

**Comment by lintangtimur** [comment]
Contoh header

GET /v3/user/public_keys HTTP/1.1
OS: Android
OS-Version: 7.1.1
App-Version: 3.32.1
client-id: ovo_android
device-id: 87586011-0127-3469-aa2a-6016a68d45f6
Host: agw.ovo.id
Connection: close
Accept-Encoding: gzip, deflate
User-Agent: okhttp/4.9.0

**Reply by g00m00nry0ng** [reply]
barusan coba, response nya null dan sms OTP yg masuk tidak berupa angka melainkan berupa link. 2x hit selalu seperti ini, dan begitu yg ketiga dapat response : "{\"data\":null,\"response_code\":\"OV00015\",\"response_message\":\"Sabar ya, permintaan kode udah diproses.\"}"

setelah menunggu beberapa saat dan hit lagi, namun tetap berupa link seperti di awal.

**Reply by lintangtimur** [reply]
berupa link karena login di device baru, ada cooldown 60 detik sesudah permintaan OTP

**Reply by cvkjm** [reply]
maaf mau tanya setelah dapat public keys rsa lalu bgmn ya request nya? soalnya saya coba2 bad request melulu

**Reply by lintangtimur** [reply]
ada endpoint lagi, user request ke server, ada parameter `password` tapi di encrypt dengan format RSA.
perlu di reverse APK nya isi dari value passwordnya. Mungkin value dari public keys nya untuk encrypt value nya.

**Comment by cvkjm** [comment]
Contoh requestnya gimana tu gan keluar link gitu?

**Reply by Willi8910** [reply]
Bodynya gini gan
Headernya saya sesuaikan seperti get public key di atas

{
  "channel_code": "ovo_android",
  "device_id": "87586011-0127-3469-aa2a-6016a68d4500",
  "msisdn": "+62896xxxxxxxx",
  "otp": {
    "locale": "ID",
    "sms_hash": "m9mj4ctIVR8"
  }
}

**Reply by g00m00nry0ng** [reply]
Ada yg sudah coba apakah dengan link ini bisa lanjut ke tahap berikutnya dan mendapatkan token seperti sebelumnya?

**Reply by Willi8910** [reply]
Saya sudah bisa gan tapi otpnya diganti jadi request part 'code' yg dikirim melalui sms
Jadi seperti ini requestnya

{
  "channel_code": "ovo_android",
  "device_id": "87586011-0127-3469-aa2a-6016a68d4500",
  "msisdn": "+6289xxxx",
  "otp": {
    "otp": "ef479da7d0ec0eb51b55ae900f90e2826a1e4620840e86a0faa8215e09cea615",
    "otp_ref_id": "0577b2ee-9314-4383-b0cd-b2a23a9e91d0",
    "type": "LOGIN"
  }
}

**Reply by g00m00nry0ng** [reply]
@Willi8910 ini step 2 ya Gan? saya masih mentok step 1 dimana sms yg masuk selalu link, dan request yg dikirim tidak ada response/null. btw device_id ini dari mana ya?

edited, device id sudah menggunakan gen_uuid, tapi masih no response/null.

**Reply by Willi8910** [reply]
Iya gan di step 2, untuk device id di step 2 ini harus sama dengan device id di step 1 gan,
Kalau masih testing pake 1 device id aja yg sama semua dulu kecuali sudah di ban dari ovonya, baru diganti karakter random untuk device id baru

**Reply by g00m00nry0ng** [reply]
Link yg dikirim ke sms, agan bisa buka di app OVO kah? di saya ga bisa dibuka dideteksi beda ponsel.

**Reply by Willi8910** [reply]
Iya gan tidak bisa juga, harus verifikasi pakai device id yg sama

**Reply by g00m00nry0ng** [reply]
Trus lanjut ke step 2 gimana ya. Soalnya step 1 selalu kirim link, dan kalau ga bisa di verifikasi di device yg sama, ga akan ngirim otp angka yg dibutuhkan buat step 2.

**Reply by Willi8910** [reply]
> Saya sudah bisa gan tapi otpnya diganti jadi request part 'code' yg dikirim melalui sms
> Jadi seperti ini requestnya
> 
> {
> "channel_code": "ovo_android",
> "device_id": "87586011-0127-3469-aa2a-6016a68d4500",
> "msisdn": "+6289xxxx",
> "otp": {
> "otp": "ef479da7d0ec0eb51b55ae900f90e2826a1e4620840e86a0faa8215e09cea615",
> "otp_ref_id": "0577b2ee-9314-4383-b0cd-b2a23a9e91d0",
> "type": "LOGIN"
> }
> }

Di step 2 gini gan punya saya, otpnya itu berupa code link dari sms yg diterima

**Reply by g00m00nry0ng** [reply]
@Willi8910 mantab. ane juga coba sukses step 2. makasih gan. jadi udah sukses sampe dapet authetication token ya gan?

**Reply by Willi8910** [reply]
Belum gan, masih stuck juga di step 4

**Comment by lintangtimur** [comment]
kalo login di device baru, akan dapet link, link itu dibuka yang ada app ovo nya.
Trus akses ke endpoint public key, lanjut ke step 4.

**Reply by g00m00nry0ng** [reply]
Pada saat buka link di hp yg ada app OVO, ada waning "Link tidak berlaku" itu diabaikan saja ya?

**Reply by lintangtimur** [reply]
Pas saya test itu haruse valid berlaku.

**Reply by g00m00nry0ng** [reply]
Di saya terdeteksi hp yg membuka link berbeda dengan yg request. Tertulis "Kamu hanya bisa melakukan verifikasi dengan mengeklik link via SMS dari ponsel yang sama". 

Device ID didapat dari mana ya untuk hp yg digunakan supaya ga begini?

**Reply by Willi8910** [reply]
gan, di step 4 value dari passwordnya dapat dari mana ya? Saya coba pakai dari public key tidak bisa gan

**Reply by lintangtimur** [reply]
> Di saya terdeteksi hp yg membuka link berbeda dengan yg request. Tertulis "Kamu hanya bisa melakukan verifikasi dengan mengeklik link via SMS dari ponsel yang sama".
> 
> Device ID didapat dari mana ya untuk hp yg digunakan supaya ga begini?

device ID generate sendiri, UUIDv4

> gan, di step 4 value dari passwordnya dapat dari mana ya? Saya coba pakai dari public key tidak bisa gan

kemungkinan pakai ini #54

**Reply by Willi8910** [reply]
Oke thanks gan

**Reply by g00m00nry0ng** [reply]
--edited--

clear

**Reply by lintangtimur** [reply]
respon dari mana?

**Reply by g00m00nry0ng** [reply]
step 1 gan. yg kalau hit 2x berturut-turut akan mendapat reponse : {"data":null,"response_code":"OV00015","response_message":"Sabar ya, permintaan kode udah diproses."}

Nah hit pertama kali yg menghasilkan link otp ke sms. responsenya seperti itu

**Reply by g00m00nry0ng** [reply]
@lintangtimur maaf gan, yg response aneh abaikan saja. Sepertinya ada masalah dengan Server saya tadi coba di tempat lain response aman.

Tinggal problem gimana verifikasi link nya ini karena sudah beberapa kali coba ga bisa karena dianggap beda device_id sepertinya 

--edited--

step 1 sudah aman menggunakan cara agan @Willi8910 sekarang tinggal step 4 belum sukses di bagian password

**Reply by skyyynewbies** [reply]
@g00m00nry0ng saya juga stuck di step 4, masnya udah solved?
resultku
{
    "data": null,
    "response_code": "OV00002",
    "response_message": "Bad Request: Invalid decrypted string for action: LOGIN"
}

**Reply by skyyynewbies** [reply]
@Willi8910 punya mas udah solved step 4 nya?

**Reply by g00m00nry0ng** [reply]
> @g00m00nry0ng saya juga stuck di step 4, masnya udah solved?
> resultku
> {
> "data": null,
> "response_code": "OV00002",
> "response_message": "Bad Request: Invalid decrypted string for action: LOGIN"
> }

Belum gan, masih stuck juga di step 4

**Comment by g00m00nry0ng** [comment]
Barusan coba step 4 dengan password yg sama seperti contoh di atas dapat respon begini

{
    "data": null,
    "response_code": "OV00539",
    "response_message": "PIN payload validation failed"
}

sepertinya password dibentuk salah satu komponennya dari PIN.

**Reply by lintangtimur** [reply]
coba di cek decompile APK nya, trus di cek parameternya

**Reply by g00m00nry0ng** [reply]
udah coba baca-baca isi decompile APK, tapi masih ga sampe ilmu saya gan. keyword RSA juga ada beberapa salah satunya yg agan post di #54

**Comment by skyyynewbies** [comment]
udah bisa tapi stuck di step 4.
{
    "data": null,
    "response_code": "OV00002",
    "response_message": "Bad Request: Invalid decrypted string for action: LOGIN"
}

**Reply by lintangtimur** [reply]
#54

**Comment by ii64** [comment]
Ada update terbaru tentang ini?

Gw ada ide supaya ga terlalu banyak reverse + implementasi
Pake gRPC aja, jadi gateway client
Barusan aja buat, pas cek login2FA ternyata error haha, jd mampir kesini, ternyata udh lama, pernah buat yg py jg jadi spontan biar spec nya sama, pake gRPC

bisa lihat2 di sini
https://github.com/ii64/go-ovo

**Reply by ii64** [reply]
![image](https://user-images.githubusercontent.com/26342994/125064651-0c138e80-e0db-11eb-8da7-2f8344499edc.png)

**Reply by ii64** [reply]
https://github.com/ii64/go-ovo/blob/main/proto/service.proto

**Reply by ii64** [reply]
buat group diskusi di Discord / Telegram / Slack mungkin supaya bisa notice update, faster ovo bisa lihat kita, tapi kita harus reverse *haha*

**Reply by ii64** [reply]
ii64#6077

**Reply by lintangtimur** [reply]
kenapa dibuat grpc? bukannya tetap harus di reverse ya, untuk dapat param dan responsenya.
Added discord

**Reply by ii64** [reply]
Yea atleast, autogenerated dari IDL, portable ke languages tinggal pake grpc gateway client

**Comment by SilentX26** [comment, marked as answer]
### Invalid Decrypted String / PIN payload validation failed

Untuk yg stuck di step 4 dan mengalami error sperti diatas, error tsb terjadi karna parameter value pada key password tidak valid. 

Parameter tsb seharusnya adalah string payload yang di enkripsi menggunakan public key yang di dapatkan dari **user/public_keys**, lalu di encoding menggunakan base64. 

Flow lengkapnya dibawah ini:
1. data yang di enkripsi berupa string dengan pola berikut: _**"LOGIN|" + securityCode + "|" + currentTimeMillies + "|" + deviceId + "|" + phoneNumber + "|" + deviceId + "|" + otpRefId**_. 
2.  Keterangan: 
    - **securityCode**: berupa PIN akun ovo anda
    - **currentTimeMillies**: berupa waktu (dalam milisecond) pada saat anda hit ke API OVO
    - **deviceId**: berupa device id anda
    - **phoneNumber**: berupa nomor HP anda
    - **otpRefId**: berupa ref id otp yang didapat pada saat request OTP
2. Contoh: **"LOGIN|123456|1628222553987|abcdefgh|+628xxxxxxx|abcdefgh|xxxxxxxxxxx"**
3. data di enkripsi menggunakan public key yang didapat dari **user/public_keys**
3. di encoding menggunakan base64

**Reply by lintangtimur** [reply]
mantap nih, next saya coba. Sip mas

**Reply by skyyynewbies** [reply]
jadi contoh requestnya gmana ya mas? masih belum paham

**Reply by SilentX26** [reply]
> jadi contoh requestnya gmana ya mas? masih belum paham

untuk request nya sama saja gan seperti yang diberikan diatas (pada step 4). Yang saya jabarkan diatas itu flow/algoritma untuk men-enkrpsi payload security code/pin yang digunakan pada parameter value di key password. 

jadi request nya bisa seperti ini:
```
{
   "channel_code":"ovo_android",
   "credentials":{
      "otp_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2RlaGFzaCI6IllqWXpZek15WmpFeFpqVXlOREEzWm1FeU1HVTJNakJsWWpFME56QXlZVFUiLCJyYW5kb20iOiJNalEwTWpVeU5qTTUiLCJ2ZXJzaW9uIjoxfQ.ap8bYmC05673QA2dTmhHw7L4SLsfpEXb_VG1c53Sxxxxxx",
      "password":{
         "format":"rsa",
         "value":"VALUE PARAMETER INI NANTINYA DIISI DENGAN PAYLOAD PIN YANG TELAH TER-ENKRIPSI\n"
      }
   },
   "device_id":"87586011-0127-3469xxxxxxxxxx",
   "msisdn":"+6282xxxxxxxx",
   "push_notification_id":"fs-DYcGaRbKERLhF4hkQ92:APA91bEjjUFzzFvadIKtdrrqsyrGH26xLRR5-Oyym2l9Ybv0O1cnvqA14ghuTbXz0ogazN-Kw6iGxW2klakANBaVXoFCLrT4hWJJ5FCGOz2o5bGE7RX6XpxndNkcxnqpWat449vBvYSa"
}

```

**Reply by skyyynewbies** [reply]
> > jadi contoh requestnya gmana ya mas? masih belum paham
> 
> untuk request nya sama saja gan seperti yang diberikan diatas (pada step 4). Yang saya jabarkan diatas itu flow/algoritma untuk men-enkrpsi payload security code/pin yang digunakan pada parameter value di key password.
> 
> jadi request nya bisa seperti ini:
> 
> ```
> {
>    "channel_code":"ovo_android",
>    "credentials":{
>       "otp_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2RlaGFzaCI6IllqWXpZek15WmpFeFpqVXlOREEzWm1FeU1HVTJNakJsWWpFME56QXlZVFUiLCJyYW5kb20iOiJNalEwTWpVeU5qTTUiLCJ2ZXJzaW9uIjoxfQ.ap8bYmC05673QA2dTmhHw7L4SLsfpEXb_VG1c53Sxxxxxx",
>       "password":{
>          "format":"rsa",
>          "value":"VALUE PARAMETER INI NANTINYA DIISI DENGAN PAYLOAD PIN YANG TELAH TER-ENKRIPSI\n"
>       }
>    },
>    "device_id":"87586011-0127-3469xxxxxxxxxx",
>    "msisdn":"+6282xxxxxxxx",
>    "push_notification_id":"fs-DYcGaRbKERLhF4hkQ92:APA91bEjjUFzzFvadIKtdrrqsyrGH26xLRR5-Oyym2l9Ybv0O1cnvqA14ghuTbXz0ogazN-Kw6iGxW2klakANBaVXoFCLrT4hWJJ5FCGOz2o5bGE7RX6XpxndNkcxnqpWat449vBvYSa"
> }
> ```

**_currentTimeMillies_** itu waktu pas hit api yg step 1 atau step 4 mas?

**Reply by SilentX26** [reply]
Step 4 gan

**Reply by skyyynewbies** [reply]
lalu utk enkripsinya gimana ya?

LOGIN|250302|1628248886141|c6be07144a9a2316|+62812xxxxxxxx|bf3ca2a8-977d-4bbc-9355-75ea59a3924f

enkripsinya pakai public key yg didapat di step 3 kan?

resultnya: **_Bad Request: Invalid decrypted string for action: LOGIN_**

**Reply by SilentX26** [reply]
**_"LOGIN|" + securityCode + "|" + currentTimeMillies + "|" + deviceId + "|" + phoneNumber + "|" + deviceId + "|" + otpRefId_**
perhatikan lagi pola string nya gan, setelah phone number, ada lagi device id, baru setelah itu otpRefId nya

**Reply by skyyynewbies** [reply]
> **_"LOGIN|" + securityCode + "|" + currentTimeMillies + "|" + deviceId + "|" + phoneNumber + "|" + deviceId + "|" + otpRefId_**
> perhatikan lagi pola string nya gan, setelah phone number, ada lagi device id, baru setelah itu otpRefId nya

Ok sudah bisa mas

```
{
  "response_code": "OV00000",
  "response_message": "Success",
  "data": {
    "auth": {
      "access_token": "eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpxxxxxx",
      "token_type": "Bearer",
      "expires_in": 1628348698,
      "refresh_token": "eyJhbGciOiJSUzI1NiJ9.eyJleHBpcnlJbk1pbGxpxxxxxx",
      "scope": "user"
    },
    "account": {
      "ovo_id": "000104xxxxxxxxxx",
      "msisdn": "0812xxxxxxxx",
      "account_status": "REGISTERED",
      "kyc_status": "",
      "authentication": "PROVIDER_OTP_PASSWORD_ENCRYPTED",
      "push_notification_id": "fs-DYcGaRbKERLxxxxxxxxxxxxx"
    }
  }
}
```

Terimakasih mas @SilentX26

**Reply by SilentX26** [reply]
sama sama gan

**Reply by skyyynewbies** [reply]
> sama sama gan

Ohiya mas mau tanya lagi, itu kan ada data _**expires_in**_ ya, tipedata nya timestamp, setelah ku cek resultnya ini
`string(19) "2021-08-07 22:04:58"`
Apa **_access_token_** nya hanya berlaku selama 1 hari aja mas?

**Reply by SilentX26** [reply]
Sepertinya tidak begitu, karena saya lihat dari base64 decode nya access token (pada bagian payload) itu terdapat json sperti berikut:
{
    "expiryInMilliSeconds":604800000,
    "createTime":1628267953208,
    "secret":"sKYHjOevW/6q/tMCjnaCsFnn444SuDiaD0nf9791IxY6s3La8ZU9faWFNpr+T2M/LFdaFBLCXsfhhYQCk9UmWG8QgEXXI2wJVSipXvcSCi4SmBozUvOt+j5hEo9g+vCFuqhB/3+gAigOsfGQgmpd5vOLiPkPrHP49018gcV8/iCPbFBwvy6xOafu+GIIYyWPhW7Mxy9agblEeg5QJZO/TKYMOwaEiT+ktjJNafVygyaaxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
}

dimana _createTime_ ialah millisecond disaat access token dibuat, dan _expiryInMilliSeconds_ adalah masa access token tersebut valid (dalam millisecond), jika dikonversi ke jam akan didapat 168 jam (7 hari). Sepertinya access token akan valid selama itu.
Karena saya sudah test login ovo dari hari jumat, sampai saat ini access token saya masih valid

**Reply by alvietomhf** [reply]
Cara pake refresh token buat generate accces token baru gimana ya mas?

**Reply by SilentX26** [reply]
> Cara pake refresh token buat generate accces token baru gimana ya mas?

Nah untuk pake refresh token nya ini saya juga masih nyari" Di decompile aplikasinya, tpi belum ketemu..

**Reply by lintangtimur** [reply]
{
    "data": null,
    "response_code": "OV00539",
    "response_message": "PIN payload validation failed"
}

masih error, padahal payload udah bener
```
LOGIN|XXXXXX|1628222553987|87586011-0127-3469-aa2a-6016a68d4XXX|+6285156714xxx|87586011-0127-3469-aa2a-6016a68d4XXX|7f0e4c34-6613-48d9-bd8f-4d9dd6b91afb
```

**Reply by herihermwn** [reply]
@lintangtimur @SilentX26 untuk value "push_notification_id" itu di dapat darimana ya ?

**Reply by lintangtimur** [reply]
> @lintangtimur @SilentX26 untuk value "push_notification_id" itu di dapat darimana ya ?

hasil sniffing ovo, itu firebase kan. Masih hardcode sih

**Reply by herihermwn** [reply]
> > @lintangtimur @SilentX26 untuk value "push_notification_id" itu di dapat darimana ya ?
> 
> hasil sniffing ovo, itu firebase kan. Masih hardcode sih

Owalahh, berarti bisa di isi apa aja ?

**Reply by lintangtimur** [reply]
mungkin iya, dicoba aja

**Reply by herihermwn** [reply]
> mungkin iya, dicoba aja

Wokee, makasih gan

**Reply by herihermwn** [reply]
@skyyynewbies @SilentX26 

Boleh minta tolong ga gan ?
2 hari ini saya coba", responsenya masih aja selalu

{
    "data": null,
    "response_code": "OV00002",
    "response_message": "Bad Request: Invalid decrypted string"
}

Source code punya agan gimana ya ? untuk bagian encryptnya.

Punya saya kaya gini (Python) :

    password = "LOGIN|{}|{}|{}|{}|{}|{}".format(security_code, current_millis(), device_id, phone, device_id, otp_ref_id)

    keyPub = RSA.importKey(public_key)
    encryptor = PKCS1_OAEP.new(keyPub)
    encrypted = encryptor.encrypt(password.encode('ascii'))

    return base64.b64encode(encrypted).decode()

Terima kasih

**Reply by SilentX26** [reply]
> @skyyynewbies @SilentX26
> 
> Boleh minta tolong ga gan ?
> 2 hari ini saya coba", responsenya masih aja selalu
> 
> {
> "data": null,
> "response_code": "OV00002",
> "response_message": "Bad Request: Invalid decrypted string"
> }
> 
> Source code punya agan gimana ya ? untuk bagian encryptnya.
> 
> Punya saya kaya gini (Python) :
> 
> ```
> password = "LOGIN|{}|{}|{}|{}|{}|{}".format(security_code, current_millis(), device_id, phone, device_id, otp_ref_id)
> 
> keyPub = RSA.importKey(public_key)
> encryptor = PKCS1_OAEP.new(keyPub)
> encrypted = encryptor.encrypt(password.encode('ascii'))
> 
> return base64.b64encode(encrypted).decode()
> ```
> 
> Terima kasih

punya saya begini gan untuk step ke 4 nya, pake PHP:
barusan saya test masih aman akses token berhasil didapatkan.

![carbon](https://user-images.githubusercontent.com/55566609/129647548-1caf0e99-64a5-47ae-a379-da0b8892413d.png)

**Reply by SilentX26** [reply]
> {
> "data": null,
> "response_code": "OV00539",
> "response_message": "PIN payload validation failed"
> }
> 
> masih error, padahal payload udah bener
> 
> ```
> LOGIN|XXXXXX|1628222553987|87586011-0127-3469-aa2a-6016a68d4XXX|+6285156714xxx|87586011-0127-3469-aa2a-6016a68d4XXX|7f0e4c34-6613-48d9-bd8f-4d9dd6b91afb
> ```

mungkin salah dibagian enkripsi nya gan.

**Reply by herihermwn** [reply]
> > {
> > "data": null,
> > "response_code": "OV00539",
> > "response_message": "PIN payload validation failed"
> > }
> > masih error, padahal payload udah bener
> > ```
> > LOGIN|XXXXXX|1628222553987|87586011-0127-3469-aa2a-6016a68d4XXX|+6285156714xxx|87586011-0127-3469-aa2a-6016a68d4XXX|7f0e4c34-6613-48d9-bd8f-4d9dd6b91afb
> > ```
> 
> mungkin salah dibagian enkripsi nya gan.

Oke gan, saya coba lagi

**Reply by herihermwn** [reply]
Untuk python code enkripsi sudah bisa dan berhasil ketika saya test .

Codenya seperti berikut : 
![carbon (1)](https://user-images.githubusercontent.com/57160005/130384292-60d1ef5b-749d-4aff-b9cd-b219f7853461.png)

Semoga bisa membantu yang lain ^_^

**Reply by lintangtimur** [reply]
> Untuk python code enkripsi sudah bisa dan berhasil ketika saya test .
> 
> Codenya seperti berikut :
> ![carbon](https://user-images.githubusercontent.com/57160005/129760456-573ff7a7-9936-4921-ab37-ebe0744b6a61.png)
> 
> Semoga bisa membantu yang lain ^_^

bisa di PR ke @ii64

**Comment by lintangtimur** [comment]
+"response_message": "Anda Tidak Memiliki Akses"
masih ini di step 4, coba kirim contoh full header + format telpon

**Reply by herihermwn** [reply]
> +"response_message": "Anda Tidak Memiliki Akses"
> masih ini di step 4, coba kirim contoh full header + format telpon

Gini gan, 

Headers: 
{
    'OS': 'Android',
    'OS-Version': '9.0',
    'App-Version': '3.32.1',
    'client-id': 'ovo_android',
    'device-id': device_id,
    'Host': 'agw.ovo.id',
    'Connection': 'close',
    'Accept-Encoding': 'gzip, deflate',
    'User-Agent': 'okhttp/4.9.0',
    'Content-Type': 'application/json'
}

Format telpon:
+628588xxxxxxx

**Comment by lintangtimur** [comment]
https://github.com/lintangtimur/ovoid/discussions/56#discussion-3538536

**Comment by makanSukros** [comment]
Punten, apakah di sini ada yang mengalami hal yang sama, untuk endpoint get list mutasi (wallet/v2/transaction) mendapatkan return error seperti ini

{
error: { message: 'Transaction Not Found', code: 'OV00009' },
status: 404,
message: 'Data Not Found!'
}

**Reply by makanSukros** [reply]
> Punten, apakah di sini ada yang mengalami hal yang sama, untuk endpoint get list mutasi (wallet/v2/transaction) mendapatkan return error seperti ini
> 
> { error: { message: 'Transaction Not Found', code: 'OV00009' }, status: 404, message: 'Data Not Found!' }

@lintangtimur @ii64

**Reply by lintangtimur** [reply]
Hi,
Ini saya barusan test di versi OVO terbaru 3.73.1 tidak ada hasil capture dari endpoint tersebut
yang saya temukan 
GET `https://agw.ovo.id/payment/orders/v1/list?page=1&limit=15&productType=001`

Header yang harus ada:
`Authorization bearer token` dan `Client-Id:ovo_android`

**Reply by Juandafa055** [reply]
halo @lintangtimur ada contacc yanh bisa di hub?
